Content.Client/Xenoarchaeology/Ui/AnalysisConsoleMenu.xaml.cs

```diff
@@ -15,169 +24,226 @@ namespace Content.Client.Xenoarchaeology.Ui;
-        ProgressBar.Value = Math.Clamp(1.0f - (float) remaining.Divide(total), 0.0f, 1.0f);
+        _nodeExtractionValues.Sort((x, y) => x.Item2.CompareTo(y.Item2));
     }
 
-    public void SetButtonsDisabled(AnalysisConsoleUpdateState state)
+    protected override void FrameUpdate(FrameEventArgs args)
     {
-        ScanButton.Disabled = !state.CanScan;
-        PrintButton.Disabled = !state.CanPrint;
-        if (state.IsTraversalDown)
-            DownBiasButton.Pressed = true;
-        else
-            UpBiasButton.Pressed = true;
+        base.FrameUpdate(args);
+
+        if (_nextExtractStringTime == null)
+            return;
+
+        if (_timing.CurTime < _nextExtractStringTime)
+            return;
 
-        ExtractButton.Disabled = false;
-        if (!state.ServerConnected)
+        if (_nodeExtractionValues.Count == 0)
         {
-            ExtractButton.Disabled = true;
-            ExtractButton.ToolTip = Loc.GetString("analysis-console-no-server-connected");
+            ExtractContainer.Visible = false;
+            NodeViewContainer.Visible = true;
+            _nextExtractStringTime = null;
+            return;
         }
-        else if (!state.CanScan)
+
+        var (message, value) = _nodeExtractionValues.First();
+        _nodeExtractionValues.RemoveAt(0);
+
+        var delay = _nodeExtractionValues.Count == 0 ? TimeSpan.FromSeconds(3) : TimeSpan.FromSeconds(0.25);
+        _nextExtractStringTime = _timing.CurTime + delay;
+        _extractionSum += value;
+        _extractionMessage.AddMarkupOrThrow(message);
+        _extractionMessage.PushNewline();
+
+        ExtractionResearchLabel.SetMessage(_extractionMessage);
+        ExtractionSumLabel.SetMarkup(Loc.GetString("analysis-console-extract-sum", ("value", _extractionSum)));
+
+        if (_playerManager.LocalSession?.AttachedEntity is { } attachedEntity)
         {
-            ExtractButton.Disabled = true;
-
-            // CanScan can be false if either there's no analyzer connected or if there's
-            // no entity on the scanner. The `Information` text will always tell the user
-            // of the former case, but in the latter, it'll only show a message if a scan
-            // has never been performed, so add a tooltip to indicate that the artifact
-            // is gone.
-            if (state.AnalyzerConnected)
-            {
-                ExtractButton.ToolTip = Loc.GetString("analysis-console-no-artifact-placed");
-            }
-            else
-            {
-                ExtractButton.ToolTip = null;
-            }
+            var volume = _nodeExtractionValues.Count == 0 ? 1f : -10f;
+            _audio.PlayGlobal(_owner.Comp.ScanFinishedSound, attachedEntity, AudioParams.Default.WithVolume(volume));
         }
-        else if (state.PointAmount <= 0)
+
+        if (_nodeExtractionValues.Count == 0)
+            OnExtractButtonPressed?.Invoke();
+    }
+
+    public void Update(Entity<AnalysisConsoleComponent> ent)
```

Comment:
is called when Console entity becomes dirty and gets synced. Just some boring stuff with labels...

Reply:
And actually there is a problem with it - console is not becoming dirty too often as its fields are not containing anything specific. Durability change does not require sync, for example - this looks like a mistake. Same way list of node requirements/trigger does not get updated